File: lingting-core/src/main/java/live/lingting/framework/value/cycle/StepCycleValue.java
package live.lingting.framework.value.cycle;

import live.lingting.framework.value.StepValue;

/**
 * @author lingting 2024-01-23 15:22
 */
public class StepCycleValue<T> extends AbstractConcurrentCycleValue<T> {

	private final StepValue<T> step;

	public StepCycleValue(StepValue<T> step) {
		this.step = step;
	}

	@Override
	public void doReset() {
		step.reset();
	}

	@Override
	public T doNext() {
		if (!step.hasNext()) {
			step.reset();
		}
		return step.next();
	}

}

File: lingting-core/src/main/kotlin/live/lingting/framework/value/cycle/StepCycleValue.kt
package live.lingting.framework.value.cycle

import live.lingting.framework.value.StepValue

/**
 * @author lingting 2024-01-23 15:22
 */
class StepCycleValue<T>(private val step: StepValue<T>) : AbstractConcurrentCycleValue<T>() {

    override fun doReset() {
        step.reset()
    }

    override fun doNext(): T {
        if (!step.hasNext()) {
            step.reset()
        }
        return step.next()
    }
}
